common/common.go

```diff
@@ -647,3 +647,38 @@ func (c *Counter) IncrementAndGet() int64 {
 	}
 	return newID
 }
+
+// ElementProcessor defines the function signature for processing an individual element with its index.
+type ElementProcessor[E any] func(index int, element E) error
+
+// ProcessElementsByBatch takes a slice of elements and processes them in batches of `batchSize` concurrently.
+// For example, if batchSize = 10 and list has 100 elements, 10 goroutines will process 10 elements concurrently
+// in each batch. Each batch completes before the next batch begins.
+// `process` is a function called for each individual element with its index and value.
+func ProcessElementsByBatch[S ~[]E, E any](batchSize int, list S, process ElementProcessor[E]) error {
+	var errs error
+	for i, s := range Batch(list, batchSize) {
+		err := CollectErrors(len(s))
+		for j, e := range s {
+			go func(index int, element E) { err.C <- process(index, element); err.Wg.Done() }((i*batchSize)+j, e)
+		}
+		errs = AppendError(errs, err.Collect())
+	}
+	return errs
+}
```

Comment:
Okay, so this is what I *think* is appropriate for your use case.
Note: Completely untested. It compiles and it's (probably) sane)
I've omitted i because I can't see that you need it, and can't imagine how it'd make sense given it's not guaranteed to be in sequence at all. If there was some use-case for it that I'm missing, then you could use a sequence generator to get it, or something. You could also maybe do this less elegantly by using mutexes and reducing the slice, or using a producer goro.
Anyway. Enough babbling. More code:
```suggestion
// PipelineProcessor defines the function signature for processing an individual elements
type PipelineProcessor[E any] func(element E) error

// ThrottledPipeline processes a slice concurrently with a throttled number of workers
func ThrottledPipeline[S ~[]E, E any](workers int, list S, process PipelineProcessor[E]) error {
	q := make(chan E, len(list))
	for _, s := range list {
		q <- s
	}
	err := CollectErrors(len(list))
	for range workers {
		go func() {
			defer err.Wg.Done()
			for {
				select {
				case s := <-q:
					err.C <- process(s)
				default:
					return
				}
			}
		}()
	}
	return err.Collect()
}
```

Reply:
There's definitely a use case for this if it works 🔥 `Note: Completely untested` 😆 Will Save this.